examples/E2E-RAG-App/gradio_interface.py

```diff
@@ -0,0 +1,242 @@
+            print(f"Memory bank registered.")
+
+    async def load_documents(self):
+        """Load documents from the specified directory into memory bank."""
+        documents = []
+        for filename in os.listdir(self.docs_dir):
+            if filename.endswith((".txt", ".md")):
+                file_path = os.path.join(self.docs_dir, filename)
+                with open(file_path, "r", encoding="utf-8") as file:
+                    content = file.read()
+                    document = Document(
+                        document_id=filename,
+                        content=content,
+                        mime_type="text/plain",
+                        metadata={"filename": filename},
+                    )
+                    documents.append(document)
+
+        if documents:
+            self.client.memory.insert(
+                bank_id=self.memory_bank_id,
+                documents=documents,
+            )
+            print(f"Loaded {len(documents)} documents from {self.docs_dir}")
+
+    async def initialize_agent(self):
+        """Initialize the agent with model registration and configuration."""
+
+        if "1b" in MODEL_NAME:
+            model_name = "Llama3.2-1B-Instruct"
+        elif "3b" in MODEL_NAME:
+            model_name = "Llama3.2-3B-Instruct"
+        elif "8b" in MODEL_NAME:
+            model_name = "Llama3.1-8B-Instruct"
+        else:
+            model_name = MODEL_NAME
+
+        agent_config = AgentConfig(
+            model=model_name,
+            instructions="You are a helpful assistant that can answer questions based on provided documents. Return your answer short and concise, less than 50 words.",
+            sampling_params={"strategy": "greedy", "temperature": 1.0, "top_p": 0.9},
+            tools=[
+                {
+                    "type": "memory",
+                    "memory_bank_configs": [
+                        {"bank_id": self.memory_bank_id, "type": "vector"}
+                    ],
+                    "query_generator_config": {"type": "default", "sep": " "},
+                    "max_tokens_in_context": 300,
+                    "max_chunks": 5,
+                }
+            ],
+            tool_choice="auto",
+            tool_prompt_format="json",
+            enable_session_persistence=True,
+        )
+        self.agent = Agent(self.client, agent_config)
+        self.session_id = self.agent.create_session(f"session-{uuid.uuid4()}")
+
+    def chat_stream(
+        self, message: str, history: List[List[str]]
+    ) -> Generator[List[List[str]], None, None]:
+        """Stream chat responses token by token with proper history handling."""
+
+        history = history or []
+        history.append([message, ""])
+
+        output_queue = Queue()
+
+        def run_async():
+            async def async_process():
+                if self.agent is None:
+                    await self.initialize_system()
+
+                response = self.agent.create_turn(
```

Comment:
I think response is a mere Generator and not an AsyncGenerator. Maybe EventLogger is unnecessarily being an AsyncGenerator and that is causing this entire complexity around a thread. I will look into this, we need to get rid of all of this. This should look trivial like this:

```python
response = ...
for log in EventLogger().log(response): 
   yield history
```

Also I don't think "EventLogger" is the right abstraction at all (something for us to fix) -- you are reading things from the response, not logging!!

Reply:
Oh I think we can just make our Agent wrapper in our SDK to do sync CustomTool execution to make it return a Generator. https://github.com/meta-llama/llama-stack-client-python/blob/7495f517f1a35e82b78cc82378e439333dbe8641/src/llama_stack_client/lib/agents/agent.py#L79

Reply:
oh interesting that it is tool execution that makes it async. in that case it is not necessarily a bad design since tool execution _will_ block for long amounts of time. I feel like maybe we need two methods `create_turn()` and `acreate_turn()` like most of the SDKs do because depending on the threading model of the caller, one is more natural vs. another.

Reply:
adding support for async workflows or separate create_turn() and acreate_turn() methods would be a very welcome change and simplify setups like these!

for context; I was initially a little puzzled as well but had similar realization as @ashwinb. so had to come out with the current implementation to bridge async functions with gradio’s synchronous event loop since gradio doesn’t support async generators directly. To enable streaming, I had to make the chat_stream function synchronous

Reply:
https://github.com/meta-llama/llama-stack-client-python/pull/45 will make agent create_turn a sync function call. 
- apps migration: https://github.com/meta-llama/llama-stack-apps/pull/123

Reply:
fixed to reflect this new change